piq/pieapp.py

```diff
@@ -0,0 +1,205 @@
+    _weights_url = "https://web.ece.ucsb.edu/~ekta/projects/PieAPPv0.1/weights/PieAPPv0.1.pth"
+
+    def __init__(
+        self,
+        reduction: str = "mean",
+        data_range: Union[int, float] = 1.0,
+        stride: int = 27,
+        replace_pooling: bool = False
+    ) -> None:
+        super().__init__()
+        
+        # Load weights and initialize model
+        weights = torch.hub.load_state_dict_from_url(self._weights_url, progress=False)
+        # Fix small bug in original weights
+        weights['ref_score_subtract.weight'] = weights['ref_score_subtract.weight'].unsqueeze(1)
+        self.model = PieAPPModel(features=64)
+        self.model.load_state_dict(weights)
+
+        if replace_pooling:
+            self.model = self.replace_pooling(self.model)
+
+        # Disable gradients
+        for param in self.model.parameters():
+            param.requires_grad_(False)
+
+        self.data_range = data_range
+        self.reduction = reduction
+        self.stride = stride
+
+    def forward(self, prediction: torch.Tensor, target: torch.Tensor) -> torch.Tensor:
+        r"""Computation of PieAPP  between feature representations of prediction and target tensors.
+        Args:
+            prediction: Tensor with shape (H, W), (C, H, W) or (N, C, H, W).
+            target: Tensor with shape (H, W), (C, H, W) or (N, C, H, W).
+        """
+        _validate_input(input_tensors=(prediction, target), allow_5d=False, allow_negative=True)
+        prediction, target = _adjust_dimensions(input_tensors=(prediction, target))
+
+        N, C, _, _ = prediction.shape
+
+        self.model.to(device=prediction.device)
+        prediction_features, prediction_weights = self.get_features(prediction)
+        target_features, target_weights = self.get_features(target)
+
+        distances, weights = self.model.compute_difference(
+            target_features - prediction_features,
+            target_weights - prediction_weights
+        )
+
+        # Shape (N, NUM_PATCHES)
+        distances = distances.reshape(N, -1)
+        weights = weights.reshape(N, -1)
+
+        # Scale scores, then average across patches
+        loss = torch.stack([(d * w).sum() / w.sum() for d, w in zip(distances, weights)])
+
+        if self.reduction == 'none':
+            return loss
+
+        return {'mean': loss.mean,
+                'sum': loss.sum
+                }[self.reduction](dim=0)
+
+    def get_features(self, x: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]:
+        r"""
+        Args:
+            x: Tensor with shape (N, C, H, W)
+        
+        Returns:
+            features: List of features extracted from intermediate layers
+        """
+        # Rescale to [0, 255] range on which models was trained
+        x = (x / float(self.data_range)) * 255
+
+        x_patches = crop_patches(x, size=64, stride=self.stride)
+
+        features, weights = self.model(x_patches)
```

Comment:
What do you think about adding this line? May be useful when PieAPP is used as metric.
```suggestion
        with torch.no_grad():
            features, weights = self.model(x_patches)
```

Reply:
Added `torch.autograd.set_grad_enabled` context manager and additional flag in `init`. 
Some users may want to use PieAPP as a loss, so turning off all gradients isn't good idea.